### PR TITLE
Add metadata tables for `data_files` and `delete_files`

### PR DESCRIPTION
**Pull request by soumya-ghosh**
Implements metadata tables for `data_files` and `delete_files` - https://github.com/apache/iceberg-python/issues/1053

Have reused the logic of `files` to derive `data_files` and `delete_files`.
Also reused the test cases for files (`test_inspect_files`) as schema is same as `files`.

### CODE REVIEW

File: pyiceberg/table/__init__.py
Comment (ndrluis) [inline pyiceberg/table/__init__.py:4370]: What do you think about moving the file_content_type options to a constants module?

File: pyiceberg/table/__init__.py
Comment (ndrluis) [inline pyiceberg/table/__init__.py:4400]: I believe that using `dict(data_file.column_sizes or {})` will achieve the same behavior without the need for an if statement. What do you think?

File: pyiceberg/table/__init__.py
Comment (soumya-ghosh) [reply pyiceberg/table/__init__.py]: Updated signature of _files() to use `DataFileContent` Enum, had to use a list, as there are two types in case of delete files.

**Review by ndrluis — APPROVED** [review]

File: mkdocs/docs/api.md
Comment (kevinjqliu) [inline mkdocs/docs/api.md:847]: with the new markdown linter (#1118), this might not be necessary anymore

File: pyiceberg/table/__init__.py
Comment (kevinjqliu) [inline pyiceberg/table/__init__.py:4312]: nit: `data_file_filter` and also use Set instead of List

File: pyiceberg/table/__init__.py
Comment (kevinjqliu) [reply pyiceberg/table/__init__.py:4400]: nit: push this logic to the line above
```
                column_sizes = data_file.column_sizes or {}
```

Comment (kevinjqliu) [reply pyiceberg/table/__init__.py:4400]: also i wonder if the null/None behavior is due to the `.toPandas()` conversion when queried through Spark

File: tests/integration/test_inspect_table.py
Comment (kevinjqliu) [inline tests/integration/test_inspect_table.py:679]: is this to produce delete files?

**Review by kevinjqliu — COMMENTED** [review]
added a few comments

File: tests/integration/test_inspect_table.py
Comment (soumya-ghosh) [reply tests/integration/test_inspect_table.py:679]: Yes

Comment (soumya-ghosh) [reply tests/integration/test_inspect_table.py:679]: About this: https://github.com/apache/iceberg-python/pull/1066#discussion_r1740146920
I have checked in Spark shell as well, the values are `null`

File: pyiceberg/table/__init__.py
Comment (soumya-ghosh) [reply pyiceberg/table/__init__.py:4312]: What would be benefit of using a Set instead of List? `_files` is not a public function and the list also contains 2 elements at max. Are you suggesting it because class member functions are not really private, thus to prevent abuse of `_files`?

Comment (kevinjqliu) [reply pyiceberg/table/__init__.py:4312]: prefer Set since it doesnt make sense to pass in multiple of the same data file types.


https://github.com/apache/iceberg-python/pull/1066/files#diff-23e8153e0fd497a9212215bd2067068f3b56fa071770c7ef326db3d3d03cee9bR4418-R4425

File: tests/integration/test_inspect_table.py
Comment (kevinjqliu) [reply tests/integration/test_inspect_table.py:679]: do you know if this produces positional deletes, equality deletes, or both?

Comment (soumya-ghosh) [reply tests/integration/test_inspect_table.py:679]: I think Spark only produces positional delete files. Flink might produce equality delete files.

File: mkdocs/docs/api.md
Comment (soumya-ghosh) [reply mkdocs/docs/api.md:847]: Okay, removing it. I was not sure what this did in the first place. Just observed that other snippets of Info or Note have is wrapped in this tag, hence added the same.

File: pyiceberg/table/__init__.py
Comment (soumya-ghosh) [reply pyiceberg/table/__init__.py]: Method signature of `_files` updated.

File: tests/integration/test_inspect_table.py
Comment (kevinjqliu) [reply tests/integration/test_inspect_table.py:679]: hm, good point. I can't find any reference with Spark and equality delete

Comment (soumya-ghosh) [reply tests/integration/test_inspect_table.py:679]: Spark streaming might produce equality deletes, I tried with Flink, created an Iceberg table and upserted some data into it, observed that there were both positional and equality deletes files, which looked weird to me.

Comment (kevinjqliu) [reply tests/integration/test_inspect_table.py:679]: this was a nit comment btw, not blocking. we don't necessary have to test for equality delete files here

Comment (soumya-ghosh) [reply tests/integration/test_inspect_table.py:679]: Alright, other than this, other comments are resolved.

**Review by kevinjqliu — APPROVED** [review]

**Comment by soumya-ghosh** [comment]
@kevinjqliu could you execute the workflow for this? I've merged this PR into refactored __init__.py. Also wanted to check if integration tests are failing here.

**Comment by soumya-ghosh** [comment]
@sungwy @Fokko could you review this?

File: tests/integration/test_inspect_table.py
Comment (sungwy) [reply tests/integration/test_inspect_table.py:679]: > Spark only produces positional delete files. Flink might produce equality delete files.

That's my understanding as well. I think just testing for positional deletes is okay for now, and maybe we can think of ways in adding equality deletes in our test suite in the future.

**Comment by kevinjqliu** [comment]
thanks again @soumya-ghosh. I'll wait for another approval before merging.

**Review by sungwy — APPROVED** [review]

**Comment by sungwy** [comment]
Thanks for making this contribution @soumya-ghosh ! And thank you @kevinjqliu and @ndrluis for the detailed reviews!

**Pull request merged**
